### PR TITLE
refactor: improve logic to calculate bitcoin fee

### PR DESCRIPTION
**Pull request by kyranjamie**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/4194613079).

WIP

This PR improves the fee estimation. Previously, we used `coinselect` to handle utxo selection and fees. This library is outdated and doesn't consider the varying output script sizes of different address types. 

To fix this, and yield more accurate fee suggestions, a transaction size estimator is used. I copied this, and converted to TypeScript from a fork of this project https://jlopp.github.io/bitcoin-transaction-size-calculator/. This is done by using a naive "biggest first" model for selecting UTXOs, and using better size esimations to derive a fee.

I've moved to using the mempool.space API which seems to give the most realistic fee rates.

### CODE REVIEW

File: src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-calculate-max-spend.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// To know a users max spendable amount, we need to consider how many UTXOs they
```
Comment (fbwoolf) [inline src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-calculate-max-spend.ts]: Think this comment is on the wrong line here?

File: src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-calculate-max-spend.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import { useCurrentBtcAccountAddressIndexZero } from '@app/store/accounts/blockchain/bitcoin/bitcoin-account.hooks';
+
+import { BtcSizeFeeEstimator } from '../fees/tx-size-calculator';
+
+// have, in order to figure out the transaction size and the current fee rates
+// ts-unused-exports:disable-next-line
+export function useCalculateMaxSpend() {
+  const currentAccountBtcAddress = useCurrentBtcAccountAddressIndexZero();
+  const { data: utxos } = useGetUtxosByAddressQuery(currentAccountBtcAddress);
+  const { data: feeRate } = useBitcoinFeeRatesInVbytes();
```
Comment (fbwoolf) [inline src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-calculate-max-spend.ts]: Isn't it more accurate to say `..SatPerVbytes` for the fee rate? I feel like it is typ written in docs as `(in sat/vB)`

File: src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-coin-selection.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const { data: utxos } = useGetUtxosByAddressQuery(currentAccountBtcAddress);
+  const { data: feeRate } = useBitcoinFeeRatesInVbytes();
+  return (sum: Money) => {
+    if (!utxos || !feeRate) return;
+
+    return {
+      high: coinselect(
+        utxos,
+        [{ address: currentAccountBtcAddress, value: sum.amount.toNumber() }],
+        feeRate.fastestFee
```
Comment (fbwoolf) [inline src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-coin-selection.ts]: Should this naming ref a time too, seems odd to switch from a time prediction to just `fastestFee`?

File: src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-calculate-max-spend.ts (hunk 2 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-calculate-max-spend.ts:17]: Nvm, I see that is coming from the new api now.

File: src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-coin-selection.ts (hunk 3 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/family/bitcoin/hooks/use-coin-selection.ts]: Github is killing me! Nvm, I see that is coming from the new api now.

**Comment by fbwoolf** [comment]
Looks good @kyranjamie, I like both of the changes you are making here. It is much more clear how the coin selection is happening and feels much more in our control to debug if needed. Also, like the fee api change. I am getting stuff when I click on `Preview` with the current build...

Screen Shot 2023-02-15 at 12 40 27 PM

**Comment by fbwoolf** [comment]
>  I am getting stuff when I click on `Preview` with the current build...
> 
Think this was my error in the amount field not realizing what I entered. Just wasn't giving me an insufficient balance warning. Haha.

File: src/app/query/bitcoin/bitcoin-client.ts [hunk 4]
```diff
@@ -34,13 +34,32 @@ class AddressApi {
...
+      errorMsg: 'No fee estimates fetched',
+      url: `https://bitcoinfees.earn.com/api/v1/fees/recommended`,
+    });
+  }
+
+  async getFeeEstimatesFromMempoolSpaceApi(): Promise<FeeEstimateMempoolSpaceApi> {
     return fetchBitcoinData({
       errorMsg: 'No fee estimates fetched',
-      url: `${this.configuration.baseUrl}/fee-estimates`,
+      url: ` https://mempool.space/api/v1/fees/recommended`,
```
Comment (fbwoolf) [inline src/app/query/bitcoin/bitcoin-client.ts:62]: Should we be switching to this api for the other requests too? Might be better than using Blockstream?

Comment (kyranjamie) [reply src/app/query/bitcoin/bitcoin-client.ts:62]: maybe, looks pretty complete

File: src/app/pages/send/send-crypto-asset-form/form/btc/use-generate-bitcoin-raw-tx.ts [hunk 5]
```diff
@@ -4,45 +4,47 @@ import { hexToBytes } from '@stacks/common';
...
-          [{ address: values.recipient, value: btcToSat(values.amount).toNumber() }],
-          feeRate.estimates[1].feeRate
-        );
+          recipient: values.recipient,
+          amount: btcToSat(values.amount).toNumber(),
+          feeRate: feeRate.fastestFee,
+        });
+
+        // eslint-disable-next-line no-console
+        console.log('coinselect', { inputs, outputs, fee });
```
Comment (fbwoolf) [inline src/app/pages/send/send-crypto-asset-form/form/btc/use-generate-bitcoin-raw-tx.ts:47]: You meaning to keep these console.logs (on L81 too)?

**Review by fbwoolf — APPROVED** [review]
Looks great! 🚀

File: src/app/pages/send/send-crypto-asset-form/form/btc/use-generate-bitcoin-raw-tx.ts (hunk 5 above)
Comment (kyranjamie) [reply src/app/pages/send/send-crypto-asset-form/form/btc/use-generate-bitcoin-raw-tx.ts:47]: yeah can stay there for now, but we'll need to tidy our bedrooms as soon as it's launched

**Pull request merged**